<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="Methpat : A program for summarising CpG methylation patterns">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>Methpat</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/bjpop/methpat">View on GitHub</a>

          <h1 id="project_title">Methpat</h1>
          <h2 id="project_tagline">A program for summarising CpG methylation patterns</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/bjpop/methpat/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/bjpop/methpat/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h3>
<a id="introduction" class="anchor" href="#introduction" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Introduction</h3>

<p>This program summarises the resultant DNA methylation pattern data from the output of <a href="http://www.bioinformatics.babraham.ac.uk/projects/bismark/">Bismark</a> bismark_methylation_extractor. Information of the DNA methylation positions for each amplicon, DNA methylation patterns observed within each amplicon and their abundance counts are summarised into a tab delimited text file amenable for further downstream statistical analysis and visualization.</p>

<h2>
<a id="requirements" class="anchor" href="#requirements" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Requirements</h2>

<ul>
<li>Whole genome bisulfite short read sequencing data (WGBS), or Reduced Representation Bisulfite Sequencing (RRBS), or Bisulfite Amplicon Sequencing.</li>
<li>Bismark v0.12.2 and it’s dependencies (bwa v0.7.5a-r405).</li>
<li>Python 2.7 or above (but not Python 3).</li>
<li>Methpat and it’s dependencies <a href="https://github.com/bjpop/methpat">https://github.com/bjpop/methpat</a>.</li>
<li>A modern web browser such as Chrome, Firefox or Safari.</li>
</ul>

<h2>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Installation</h2>

<p>Methpat currently requires version 2.7 of Python.</p>

<p>The best way to install Methpat is to use the following command:</p>

<pre><code>pip install git+https://github.com/bjpop/methpat.git
</code></pre>

<p>This will automatically download and install the dependencies of methpat.</p>

<h2>
<a id="usage" class="anchor" href="#usage" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Usage</h2>

<pre><code>usage: methpat [-h] [--version] [--count_thresh THRESH] --amplicons
               AMPLICONS_FILE [--logfile FILENAME] [--html FILENAME]
               [--webassets {package,local,online}] [--title TITLE]
               [--filterpartial] [--min_cpg_percent PERCENT]
               BISMARK_FILE

Summarise methylation patterns in bismark output, and generate visualisation.

positional arguments:
  BISMARK_FILE          input bismark file

optional arguments:
  -h, --help            show this help message and exit
  --version             show program's version number and exit
  --count_thresh THRESH
                        only display methylation patterns with at least THRESH
                        number of matching reads, defaults to "0"
  --amplicons AMPLICONS_FILE
                        file containing amplicon information in TSV format
  --logfile FILENAME    log progress in FILENAME, defaults to "methpat.log"
  --html FILENAME       save visualisation in html FILENAME defaults to
                        "methpat.html"
  --webassets {package,local,online}
                        location of assets used by output visualisation web
                        page, defaults to "package"
  --title TITLE         title of the output visualisation page, defaults to
                        "Methylation Patterns"
  --filterpartial       Ignore reads which contain (at least one) unknown
                        methylation status
  --min_cpg_percent PERCENT
                        only consider CPG sites which occur at least PERCENT
                        of reads for an amplicon

</code></pre>

<h2>
<a id="explanation-of-command-line-arguments" class="anchor" href="#explanation-of-command-line-arguments" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Explanation of command line arguments</h2>

<ul>
<li>
<code>--version</code>: Display the version number of methpat and then exit.</li>
<li>
<code>--count_thresh THRESH</code>: Only display methylation patterns with at least THRESH number of matching reads.</li>
<li>
<code>--amplicons AMPLICONS_FILE</code>: Input file containing amplicon information in TSV format (described below).</li>
<li>
<code>--logfile FILENAME</code>: Output log progress in FILENAME.</li>
<li>
<code>--html FILENAME</code>: Save visualisation in html FILENAME.</li>
<li>
<code>--webassets {package,local,online}</code>: Location of assets used by output visualisation web page. Web assets are files needed by the HTML visualisation (javascript and CSS files). Each option has the following meaning:

<ul>
<li>
<code>package</code>: web asset files are found in the directory where methpat is installed on your computer. These files are automatically set up when methpat is installed. You do not need to do any extra steps to make this work, and it does not require a network connection, therefore it is the default behaviour for methpat. This mode is useful for viewing the HTML visualisation on the same computer where methpat is installed. It is not useful for sharing the visualisation with others or publishing on the internet as a web page.</li>
<li>
<code>local</code>: web asset files are found in the directory containing the HTML visualisation file - they must be copied there manually by the user. The web asset files can be found in the <code>methpat/data</code> sub-directory of the source code repository. This mode is useful for sharing the files with others who wish to view the HTML visualisation without an internet connection.</li>
<li>
<code>online</code>: The web assets are stored on the internet. This mode is useful for sharing the visualisation with others or publishing on the internet as a web page.</li>
</ul>
</li>
<li>
<code>--title TITLE</code>: Title of the output visualisation page, defaults to "Methylation Patterns".</li>
<li>
<code>--filterpartial</code>: Remove from consideration any reads which contain at least one CpG site with unknown methylation state.</li>
<li>
<code>--min_cpg_percent PERCENT</code>: Count the frequency of occurrences of each CpG site in a given amplicon and only include sites which occur in at least PERCENT of the total. This can be helpful for ignoring low frequency errors in the data which would otherwise show up as putative CpG sites, and may lead to many occurrences of unknown sites in other reads.</li>
</ul>

<h2>
<a id="input-amplicons-file-format" class="anchor" href="#input-amplicons-file-format" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Input amplicons file format</h2>

<p>The amplicons (or regions of interest) are defined by this tab-delimited text file. The order
of regions listed in this file will correspond to the order of display in the html file generated.
The file requires 7 columns which define:</p>

<ol>
<li>chromosome</li>
<li>start coordinate of amplicon/region of interest</li>
<li>end coordinate of amplicon/region of interest</li>
<li>name of amplicon/region of interest</li>
<li>size of amplicon/region of interest</li>
<li>length of forward PCR primer of amplicon/region of interest, number of base pairs to ignore from the start</li>
<li>length of reverse PCR primer of amplicon/region of interest, number of base pairs to ignore from the end</li>
</ol>

<p>See the file <a href="https://raw.githubusercontent.com/bjpop/methpat/master/examples/example_amplicons.txt">examples/example_amplicons.txt</a> for an example input amplicons file, which is used in the worked example below.</p>

<h2>
<a id="worked-example" class="anchor" href="#worked-example" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Worked example</h2>

<p>Run Bismark as per instructions to prepare the reference genome as outlined in the <a href="http://www.bioinformatics.babraham.ac.uk/projects/bismark/Bismark_User_Guide.pdf">bismark manual</a>.</p>

<p>Run Bismark to align raw read files (fastq) as below with no mismatches:</p>

<pre><code>bismark -n 0 -q -o &lt;nameOfOutPutDirectory&gt; --bam --ambiguous --chunkmbs 10204 --non_directional /path/to/reference/genome &lt;fastq.file&gt;
</code></pre>

<p>Run bismark methylation extractor to create an output text file containing read DNA
methylation pattern information:</p>

<pre><code>bismark_methylation_extractor -s &lt;bismarkAlignedFileName.bam&gt;
</code></pre>

<p>This creates a series of text files prefixed with the methylation orientation and reference
genome strand data was extracted from. For the purposes here, the files prefixed with CpG
are then processed with methpat.</p>

<p>For each bismark output file with a <code>bismark.txt</code> suffix, run the following command to generate the methpat output. The example below uses the data files in the <code>examples</code> directory of the methpat source package:</p>

<pre><code>methpat --amplicons example_amplicons.txt example1_bismark.txt &gt; example_methpat.tsv
</code></pre>

<p>The above example generates three output files:</p>

<ol>
<li>
<code>example_methpat.tsv</code>: summarises the read count and DNA methylation pattern information for all amplicons/regions of interest listed in the amplicons file.</li>
<li>
<code>methpat.log</code>: log information about the progress of the methpat program (useful for debugging purposes if things don't work as you expect, also records the command line that was used just in case you want to run the exact same command again).</li>
<li>
<code>methpat.html</code>: interactive HTML visualisation of the methylation patterns for each amplicon/region of interest. This is best viewed in Chrome, Firefox and Safari (but not Internet Explorer).</li>
</ol>

<p>See the file <a href="pages/methpat.html">examples/methpat.html</a> which contains the visualisation of the above worked example.</p>

<h2>
<a id="format-of-the-methylation-pattern-counts-output-text-file" class="anchor" href="#format-of-the-methylation-pattern-counts-output-text-file" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Format of the methylation pattern counts output text file</h2>

<p>One of the main outputs of methpat is a tab-delimited text file which summarises the counts for each methylation pattern over all amplicons from the input. In the example above this output file is called <code>example_methpat.tsv</code>. The file has seven columns containing the following information:</p>

<ol>
<li>
<code>amplicon ID</code>: the identifier of an amplicon from the input amplicons file.</li>
<li>
<code>chr</code>: the chromosome containing the amplicon</li>
<li>
<code>Base position start/CpG start</code>: the coordinate of the first CpG site in the amplicon.</li>
<li>
<code>Base position end/CpG end</code>: the coordinate of the last CpG site in the amplicon.</li>
<li>
<code>Methylation pattern</code>: an encoding of the methylation state of each CpG site in the amplicon. The code "1" means that the site is methylated. The code "0" means the site is unmethylated. The code "-" means that the methylation state of the site is unknown. An example is "11-011-0-" for an amplicon with nine CpG sites, four of which are methylated, three are unmethylated and two are unknown.</li>
<li>
<code>count</code>: the number of times this particular methylation pattern for the amplicon is seen in the Bismark file.</li>
<li>
<code>raw cpg sites</code>: the methylation state of each CpG site for this pattern. It is a list of comma-separated pairs in the form "(CpG coordinate, methylation state)". Coordinates with unknown methylation state are not shown in the list.</li>
</ol>

<p>See the file <a href="https://raw.githubusercontent.com/bjpop/methpat/master/examples/example_methpat.tsv">examples/methpat.tsv</a> which contains the methylation counts from the above worked example.</p>

<h2>
<a id="visualisation" class="anchor" href="#visualisation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Visualisation</h2>

<p>Methpat outputs a HTML visualisation of the methylation patterns for each amplicon in the input. By default the file is called <code>methpat.html</code> but you can override this with another name using the <code>--html</code> command line argument.</p>

<p>You can view the visualisation in a web browser (though it is probably best to use Chrome, Firefox or Safari instead of Microsoft Explorer).</p>

<p>Each amplicon in the input appears in the visualisation. The order of the amplicons in the visualisation is the same as the input file. You can dynamically re-order the amplicons from within the <em>Order</em> tab at the top of the page. </p>

<p>A screenshot is shown below for the DAPK1_p1_106 amplicon from the worked example. The red letters are annotations which have been added for the purpose of this document, they are referred to in the explanation below the image.</p>

<p><img src="images/example_methpat_vis.png" alt="Screen grab of part of the methpat visualisation for the worked example"></p>

<p>The main part of the visualisation (A) shows all the methylation patterns for the amplicon. Each pattern is displayed as a vertical column of coloured boxes. By default a yellow box indicates a methylated CpG site, a red box indicates an unmethylated CpG site, and a blue box indicates a CpG site whose methylation state is unknown. By default the patterns are sorted left-to-right in order of frequency of occurrence. The columns are sequentially numbered. The numbers are merely ordinals that can be used to refer to columns, which can be useful when discussing a diagram within a manuscript. In the image above the most frequent pattern is methylated in all CpG sites (all boxes are yellow in column number 1).</p>

<p>The coordinates of each CpG site are listed on the left side of the digram (C), and they are shown in ascending order from top-to-bottom.</p>

<p>A frequency histogram of each pattern is shown as a bar chart below the patterns (B). By default the vertical axis of the graph shows raw read counts for each pattern, however this can be changed to a percentage of total reads. The vertical scale is linear by default but it can be changed to log scale.</p>

<p>The proportion of methylated, unmethylated and unknown methylation for each CpG site across all reads for a given amplicon are shown in coloured rectangles to the left of the patterns (D). In the image above the first CpG site (90112806) is predominantly methylated (mostly yellow), with a small amount of unmethylated sites (red).</p>

<p>The green button at the bottom of the diagram (E) allows you to save a copy of the diagram to file in PNG format to your local computer. By default the saved file will be named after the amplicon with a ".png" suffix. The scale of the PNG file defaults to 1 (same size in pixels as the web page rendering) but this can be increased for larger output images and increased resolution. You may want to set this number higher if you wish to use the PNG files in a publication (downscale the higher-resolution files).</p>

<p>Many aspects of the visualisation can be changed in the "Settings" tab. A screen shot of the settings tab is shown below:</p>

<p><img src="images/methpat_vis_settings.png" alt="Screen grab of part of the methpat visualisation settings"></p>

<p>If you change any of the visualisation settings the methylation patterns in the <em>Graphs</em> tab will re-draw automatically when that tab is displayed.</p>

<p>The setting options are described below:</p>

<ul>
<li>
<p>Methylation patterns:</p>

<ul>
<li>Show methylation site spacing: check on this radio button to display the relative spacing between CpG sites. The vertical width of the spacing is controlled by <em>Methylation site spacing factor</em>, which is a multiple of the normal (non-spaced) display of the patterns. It defaults to 2 which means that the spaced display will be twice as wide as the non-spaced display.</li>
<li>Sort methylation patterns by: The ordering of the methylation patterns across the screen. The ordering can be by "epiallele frequency" (pattern frequency count) or by "degree of methylation".</li>
<li>Sort direction (left to right): The ordering of methylation patterns across the screen in left-to-right direction. The ordering can be "ascending" or "descending".</li>
<li>Pattern read threshold (percent): only display methylation patterns with at least this percentage of reads, out of all the reads for the amplicon. This allows you to dynamically filter out patterns with low relative frequency. </li>
<li>PNG file save scale factor: the scaling factor for saving the PNG file. By default the saved PNG file will have the same size (and hence number of pixels) as the rendered graph in the web browser. You can increase this scale (and hence image size and resolution) by making this number larger. This may be useful for situations where you need a higher resolution image for the purposes of inclusion in a publication.</li>
<li>Cell size (pixels): The side-length in pixels of the coloured squares used to show the methylation patterns. Default is 15 pixels.</li>
<li>Scale pattern intensity: If set to "true" the colour intensity of each methylation pattern will be scaled to indicate its frequency. More frequent patterns are shown brighter and less frequent patterns are shown darker. This is set to "false" by default.</li>
</ul>
</li>
<li>
<p>Histogram:</p>

<ul>
<li>Histogram scaling: How to display the vertical scale of the histogram. The scale can be "linear" or "log".</li>
<li>Histogram visible: Should the histogram be displayed? If set to "true" then the histogram will be displayed. If set to "false" the histogram will not be displayed.</li>
<li>Histogram height (pixels): The maximum vertical height of the histogram in pixels. Defaults to 100 pixels.</li>
<li>Histogram units: The units of the vertical scale on the histogram. The units can be "absolute" or "relative". The absolute scale shows raw counts. The relative scale shows the percentage of all reads for the amplicon.</li>
</ul>
</li>
<li>
<p>Colour:</p>

<ul>
<li>Methylated site: The colour to use for a methylated site.</li>
<li>Unmethylated site: The colour to use for an unmethylated site.</li>
<li>Unknown site: The colour to use for a CpG site with unknown methylation state.</li>
<li>Histogram: The colour of the bars in the histogram.</li>
</ul>
</li>
</ul>

<p>The relative order of the amplicons in the <em>Graphs</em> tab can be changed dynamically from within the <em>Order</em> tab. The image below shows an example for one set of amplicons. The order of the amplicon names in the list reflects the order that their methylation pattern visualisations will be shown. You can drag the amplicon names in the list to modify their ordering.</p>

<p><img src="images/methpat_order.png" alt="Screen grab of part of the methpat amplicon re-ordering tab"></p>

<p>The image below illustrates the effect of showing methylation site spacing in the visualisation. Vertical spacing between the rows of the visualisation reflects the relative genomic spacing of the sites.</p>

<p><img src="images/methpat_vis_space.png" alt="Screen grab of part of the methpat visualisation showing relative CpG site spacing"></p>
      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">Methpat maintained by <a href="https://github.com/bjpop">bjpop</a></p>
        <p>Published with <a href="https://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
