<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="Methpat : A program for summarising CpG methylation patterns">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>Methpat</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/bjpop/methpat">View on GitHub</a>

          <h1 id="project_title">Methpat</h1>
          <h2 id="project_tagline">A program for summarising CpG methylation patterns</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/bjpop/methpat/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/bjpop/methpat/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h3>
<a id="introduction" class="anchor" href="#introduction" aria-hidden="true"><span class="octicon octicon-link"></span></a>Introduction</h3>

<p>This program summarises the resultant DNA methylation pattern data from the output of <a href="http://www.bioinformatics.babraham.ac.uk/projects/bismark/">Bismark</a> bismark_methylation_extractor. Information of the DNA methylation positions for each amplicon, DNA methylation patterns observed within each amplicon and their abundance counts are summarised into a tab delimited text file amenable for further downstream statistical analysis and visualization.</p>

<h2>
<a id="requirements" class="anchor" href="#requirements" aria-hidden="true"><span class="octicon octicon-link"></span></a>Requirements</h2>

<ul>
<li>Whole genome bisulfite short read sequencing data (WGBS), or Reduced Representation Bisulfite Sequencing (RRBS), or Bisulfite Amplicon Sequencing.</li>
<li>Bismark v0.12.2 and it’s dependencies (bwa v0.7.5a-r405).</li>
<li>Python 2.7 or above (but not Python 3).</li>
<li>Methpat and it’s dependencies <a href="https://github.com/bjpop/methpat">https://github.com/bjpop/methpat</a>.</li>
<li>A modern web browser such as Chrome, Firefox or Safari.</li>
</ul>

<h2>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Installation</h2>

<p>Methpat currently requires version 2.7 of Python.</p>

<p>The best way to install Methpat is to use the following command:</p>

<pre><code>pip install git+https://github.com/bjpop/methpat.git
</code></pre>

<p>This will automatically download and install the dependencies of methpat.</p>

<h2>
<a id="usage" class="anchor" href="#usage" aria-hidden="true"><span class="octicon octicon-link"></span></a>Usage</h2>

<pre><code>usage: methpat [-h] [--dump_reads FILE] [--count_thresh THRESH] --amplicons
               AMPLICONS_FILE --logfile FILENAME --html FILENAME
               [--webassets {package,local,online}] [--title TITLE]
               BISMARK_FILE

Count methylation patterns in bismark output

positional arguments:
  BISMARK_FILE          input bismark file

optional arguments:
  -h, --help            show this help message and exit
  --dump_reads FILE     dump the read methylation information to FILE
  --count_thresh THRESH
                        only display methylation patterns with at least THRESH
                        number of matching reads
  --amplicons AMPLICONS_FILE
                        file containing amplicon information in TSV format
  --logfile FILENAME    log progress in FILENAME
  --html FILENAME       save visualisation in html FILENAME
  --webassets {package,local,online}
                        location of assets used by output visualisation web
                        page, defaults to package
  --title TITLE         title of the output visualisation page, defaults to
                        Methylation Patterns
</code></pre>

<h2>
<a id="explanation-of-arguments" class="anchor" href="#explanation-of-arguments" aria-hidden="true"><span class="octicon octicon-link"></span></a>Explanation of arguments</h2>

<ul>
<li>
<strong>--dump_reads FILE</strong>: Dump the read methylation information to FILE.</li>
<li>
<strong>--count_thresh THRESH</strong>: Only display methylation patterns with at least THRESH number of matching reads.</li>
<li>
<strong>--amplicons AMPLICONS_FILE</strong>: Input file containing amplicon information in TSV format.</li>
<li>
<strong>--logfile FILENAME</strong>: Output log progress in FILENAME.</li>
<li>
<strong>--html FILENAME</strong>: Save visualisation in html FILENAME.</li>
<li>
<strong>--webassets {package,local,online}</strong>: Location of assets used by output visualisation web page, defaults to package.</li>
<li>
<strong>--title TITLE</strong>: Title of the output visualisation page, defaults to "Methylation Patterns".</li>
</ul>

<h2>
<a id="amplicons-file-definintion" class="anchor" href="#amplicons-file-definintion" aria-hidden="true"><span class="octicon octicon-link"></span></a>Amplicons File Definintion</h2>

<p>The amplicons or regions of interest are defined by this tab-delimited text file. The order
of regions listed in this file will correspond to the order of display in the html file generated.
The file requires 7 columns which define:</p>

<ol>
<li>chromosome</li>
<li>start coordinate of amplicon/region of interest</li>
<li>end coordinate of amplicon/region of interest</li>
<li>name of amplicon/region of interest</li>
<li>size of amplicon/region of interest</li>
<li>length of forward PCR primer of amplicon/region of interest, number of base pairs to ignore from the start</li>
<li>length of reverse PCR primer of amplicon/region of interest, number of base pairs to ignore from the end</li>
</ol>

<p>See the file <code>Examples/AmpliconDetails.txt</code> for a working example.</p>

<h2>
<a id="worked-example" class="anchor" href="#worked-example" aria-hidden="true"><span class="octicon octicon-link"></span></a>Worked example</h2>

<p>Run Bismark as per instructions to prepare the reference genome as outlined in the <a href="http://www.bioinformatics.babraham.ac.uk/projects/bismark/Bismark_User_Guide.pdf">bismark manual</a>.</p>

<p>Run Bismark to align raw read files (fastq) as below with no mismatches:</p>

<pre><code>bismark -n 0 -q -o &lt;nameOfOutPutDirectory&gt; --bam --ambiguous --chunkmbs 10204 --non_directional /path/to/reference/genome &lt;fastq.file&gt;
</code></pre>

<p>Run bismark methylation extractor to create an output text file containing read DNA
methylation pattern information:</p>

<pre><code>bismark_methylation_extractor -s &lt;bismarkAlignedFileName.bam&gt;
</code></pre>

<p>This creates a series of text files prefixed with the methylation orientation and reference
genome strand data was extracted from. For the purposes here, the files prefixed with CpG
are then processed with methpat.</p>
      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">Methpat maintained by <a href="https://github.com/bjpop">bjpop</a></p>
        <p>Published with <a href="http://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
