<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="Methpat : A program for summarising CpG methylation patterns">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>Methpat</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/bjpop/methpat">View on GitHub</a>

          <h1 id="project_title">Methpat</h1>
          <h2 id="project_tagline">A program for summarising CpG methylation patterns</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/bjpop/methpat/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/bjpop/methpat/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h3>
<a id="introduction" class="anchor" href="#introduction" aria-hidden="true"><span class="octicon octicon-link"></span></a>Introduction</h3>

<p>This program summarises the resultant DNA methylation pattern data from the output of <a href="http://www.bioinformatics.babraham.ac.uk/projects/bismark/">Bismark</a> bismark_methylation_extractor. Information of the DNA methylation positions for each amplicon, DNA methylation patterns observed within each amplicon and their abundance counts are summarised into a tab delimited text file amenable for further downstream statistical analysis and visualization.</p>

<h2>
<a id="requirements" class="anchor" href="#requirements" aria-hidden="true"><span class="octicon octicon-link"></span></a>Requirements</h2>

<ul>
<li>Whole genome bisulfite short read sequencing data (WGBS), or Reduced Representation Bisulfite Sequencing (RRBS), or Bisulfite Amplicon Sequencing.</li>
<li>Bismark v0.12.2 and it’s dependencies (bwa v0.7.5a-r405).</li>
<li>Python 2.7 or above (but not Python 3).</li>
<li>Methpat and it’s dependencies <a href="https://github.com/bjpop/methpat">https://github.com/bjpop/methpat</a>.</li>
<li>A modern web browser such as Chrome, Firefox or Safari.</li>
</ul>

<h2>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Installation</h2>

<p>Methpat currently requires version 2.7 of Python.</p>

<p>The best way to install Methpat is to use the following command:</p>

<pre><code>pip install git+https://github.com/bjpop/methpat.git
</code></pre>

<p>This will automatically download and install the dependencies of methpat.</p>

<h2>
<a id="usage" class="anchor" href="#usage" aria-hidden="true"><span class="octicon octicon-link"></span></a>Usage</h2>

<pre><code>methpat [-h] [--count_thresh THRESH] --amplicons AMPLICONS_FILE
        [--logfile FILENAME] [--html FILENAME]
        [--webassets {package,local,online}] [--title TITLE]
        BISMARK_FILE

Summarise methylation patterns in bismark output, and generate visualisation.

positional arguments:
  BISMARK_FILE          input bismark file

optional arguments:
  -h, --help            show this help message and exit
  --count_thresh THRESH
                        only display methylation patterns with at least THRESH
                        number of matching reads, defaults to "0"
  --amplicons AMPLICONS_FILE
                        file containing amplicon information in TSV format
  --logfile FILENAME    log progress in FILENAME, defaults to "methpat.log"
  --html FILENAME       save visualisation in html FILENAME defaults to
                        "methpat.html"
  --webassets {package,local,online}
                        location of assets used by output visualisation web
                        page, defaults to "package"
  --title TITLE         title of the output visualisation page, defaults to
                        "Methylation Patterns"
</code></pre>

<h2>
<a id="explanation-of-arguments" class="anchor" href="#explanation-of-arguments" aria-hidden="true"><span class="octicon octicon-link"></span></a>Explanation of arguments</h2>

<ul>
<li>
<code>--dump_reads FILE</code>: Dump the read methylation information to FILE. Mainly useful for debugging methpat behaviour.</li>
<li>
<code>--count_thresh THRESH</code>: Only display methylation patterns with at least THRESH number of matching reads.</li>
<li>
<code>--amplicons AMPLICONS_FILE</code>: Input file containing amplicon information in TSV format (described below).</li>
<li>
<code>--logfile FILENAME</code>: Output log progress in FILENAME.</li>
<li>
<code>--html FILENAME</code>: Save visualisation in html FILENAME.</li>
<li>
<code>--webassets {package,local,online}</code>: Location of assets used by output visualisation web page. Web assets are files needed by the HTML visualisation (javascript and CSS files). Each option has the following meaning:

<ul>
<li>
<code>package</code>: web asset files are found in the directory where methpat is installed on your computer. These files are automatically set up when methpat is installed. You do not need to do any extra steps to make this work, therefore it is the default behaviour for methpat. This mode is useful for viewing the HTML visualisation on the same computer where methpat is installed. It is not useful for sharing the visualisation with others or publishing on the internet as a web page.</li>
<li>
<code>local</code>: web asset files are found in the directory containing the HTML visualisation file - they must be copied there manually by the user. The web asset files can be found in the <code>methpat/data</code> sub-directory of the source code repository. This mode is useful for sharing the files with others who wish to view the HTML visualisation without an internet connection.</li>
<li>
<code>online</code>: The web assets are stored on the internet. This mode is useful for sharing the visualisation with others or publishing on the internet as a web page.</li>
</ul>
</li>
<li>
<code>--title TITLE</code>: Title of the output visualisation page, defaults to "Methylation Patterns".</li>
</ul>

<h2>
<a id="amplicons-file-definintion" class="anchor" href="#amplicons-file-definintion" aria-hidden="true"><span class="octicon octicon-link"></span></a>Amplicons File Definintion</h2>

<p>The amplicons or regions of interest are defined by this tab-delimited text file. The order
of regions listed in this file will correspond to the order of display in the html file generated.
The file requires 7 columns which define:</p>

<ol>
<li>chromosome</li>
<li>start coordinate of amplicon/region of interest</li>
<li>end coordinate of amplicon/region of interest</li>
<li>name of amplicon/region of interest</li>
<li>size of amplicon/region of interest</li>
<li>length of forward PCR primer of amplicon/region of interest, number of base pairs to ignore from the start</li>
<li>length of reverse PCR primer of amplicon/region of interest, number of base pairs to ignore from the end</li>
</ol>

<p>See the file <code>examples/example_amplicons.txt</code> for a working example.</p>

<h2>
<a id="worked-example" class="anchor" href="#worked-example" aria-hidden="true"><span class="octicon octicon-link"></span></a>Worked example</h2>

<p>Run Bismark as per instructions to prepare the reference genome as outlined in the <a href="http://www.bioinformatics.babraham.ac.uk/projects/bismark/Bismark_User_Guide.pdf">bismark manual</a>.</p>

<p>Run Bismark to align raw read files (fastq) as below with no mismatches:</p>

<pre><code>bismark -n 0 -q -o &lt;nameOfOutPutDirectory&gt; --bam --ambiguous --chunkmbs 10204 --non_directional /path/to/reference/genome &lt;fastq.file&gt;
</code></pre>

<p>Run bismark methylation extractor to create an output text file containing read DNA
methylation pattern information:</p>

<pre><code>bismark_methylation_extractor -s &lt;bismarkAlignedFileName.bam&gt;
</code></pre>

<p>This creates a series of text files prefixed with the methylation orientation and reference
genome strand data was extracted from. For the purposes here, the files prefixed with CpG
are then processed with methpat.</p>

<p>For each bismark output file with a <code>bismark.txt</code> suffix, run the following command to generate the methpat output. The example below uses the data files in the <code>examples</code> directory of the methpat source package:</p>

<pre><code>methpat --amplicons example_amplicons.txt example1_bismark.txt &gt; example_methpat.tsv
</code></pre>

<p>The above example generates three output files:</p>

<ol>
<li>
<code>example_methpat.tsv</code>: summarises the read count and DNA methylation pattern information for all amplicons/regions of interest listed in the amplicons file.</li>
<li>
<code>methpat.log</code>: log information about the progress of the methpat program (useful for debugging purposes if things don't work as you expect, also records the command line that was used just in case you want to run the exact same command again).</li>
<li>
<code>methpat.html</code>: interactive HTML visualisation of the methylation patterns for each amplicon/region of interest. This is best viewed in Chrome, Firefox and Safari (but not Internet Explorer).</li>
</ol>

<p>Here is an <a href="http://htmlpreview.github.io/?https://github.com/bjpop/methpat/blob/master/examples/methpat.html">example methpat visualisation</a> which results from the command above (though it uses the online webassets option instead of the package option).</p>

<h2>
<a id="format-of-the-methylation-pattern-counts-output-text-file" class="anchor" href="#format-of-the-methylation-pattern-counts-output-text-file" aria-hidden="true"><span class="octicon octicon-link"></span></a>Format of the methylation pattern counts output text file</h2>

<p>One of the main outputs of methpat is a tab-delimited text file which summarises the counts for each methylation pattern over all amplicons from the input. In the example above this output file is called <code>example_methpat.txt</code>. The file has seven columns containing the following information:</p>

<ol>
<li>
<code>amplicon ID</code>: the identifier of an amplicon from the input amplicons file.</li>
<li>
<code>chr</code>: the chromosome containing the amplicon</li>
<li>
<code>Base position start/CpG start</code>: the coordinate of the first CpG site in the amplicon.</li>
<li>
<code>Base position end/CpG end</code>: the coordinate of the last CpG site in the amplicon.</li>
<li>
<code>Methylation pattern</code>: an encoding of the methylation status of each CpG site in the amplicon. The code "1" means that the site is methylated. The code "0" means the site is unmethylated. The code "-" means that the methylation status of the site is unknown. An example is "11-011-0-" for an amplicon with nine CpG sites, four of which are methylated, three are unmethylated and two are unknown.</li>
<li>
<code>count</code>: the number of times this particular methylation pattern for the amplicon is seen in the Bismark file.</li>
<li>
<code>raw cpg sites</code>: the methylation status of each CpG site for this pattern. It is a list of comma-separated pairs in the form "(CpG coordinate, methylation status)". Coordinates with unknown methylation status are not shown in the list.</li>
</ol>

<p>Here is an <a href="https://raw.githubusercontent.com/bjpop/methpat/master/examples/example_methpat.txt">example methylation pattern counts file</a> resulting from the example above.</p>
      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">Methpat maintained by <a href="https://github.com/bjpop">bjpop</a></p>
        <p>Published with <a href="http://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
